LibreHardwareMonitorLib/Hardware/Controller/Corsair/CommanderProDevice.cs

```diff
@@ -0,0 +1,100 @@
+﻿using HidSharp;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace LibreHardwareMonitor.Hardware.Controller.Corsair
+{
+    internal delegate void CommanderProAction(HidStream stream, byte[] readBuffer, byte[] writeBuffer);
+
+    internal class CommanderProDevice : IDisposable
+    {
+        private readonly TimeSpan _defaultTimeout = TimeSpan.FromSeconds(3);
+        private readonly HidStream _hidStream;
+        private byte[] _writeBuffer = new byte[64];
+        private byte[] _readBuffer = new byte[16];
+        private SemaphoreSlim _semaphoreSlim = new SemaphoreSlim(1, 1);
+
+        public static bool TryOpen(HidDevice hidDevice, out CommanderProDevice stream)
+        {
+            stream = null;
+            if (hidDevice.TryOpen(out HidStream deviceStream))
+            {
+                deviceStream.ReadTimeout = 5000;
+                stream = new CommanderProDevice(deviceStream);
+                return true;
+            }
+
+            return false;
+        }
+
+        private CommanderProDevice(HidStream hidStream)
+        {
+            _hidStream = hidStream;
+        }
+
+        public bool IsOpen => _hidStream.CanRead;
+
+        public async Task DoAsync(CommanderProAction action)
+        {
+            await _semaphoreSlim.WaitAsync(_defaultTimeout);
```

Comment:
I don't know exactly why but this waits forever.

Reply:
Added a try finally loop for safer semaphore use, otherwise I don't quite know yet where it is locked?

Just put a breakpoint on every Wait on that semaphore and see who owns it.